### PR TITLE
Finish subject

### PR DESCRIPTION
**Pull request by lethihaivan**
## Related Tickets
-  [#TicketID](https://edu-redmine.sun-asterisk.vn/issues/35013)

## WHAT
- As a trainee, I can finish a subject

## Evidence (Screenshot or Video)
Screenshot from 2021-04-16 09-00-34

### CODE REVIEW

**Comment by lethihaivan** [comment]
ready

File: app/controllers/trainee/user_subjects_controller.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    UserSubject.transaction do
+      user_subject.finished!
+      user_subject.user_tasks.each do |user_task|
+        user_task.finished!
+        user_task.update! finish_at: Time.current
+      end
+    end
+    flash[:success] = t "user_subjects.finish_success"
+  rescue StandardError
+    flash[:danger] = t "user_subjects.finish_fail"
```
Comment (vanvtt-0952) [inline app/controllers/trainee/user_subjects_controller.rb:37]: vào flash rồi ko đi đâu ah bé :v

File: config/routes.rb [hunk 2]
```diff
@@ -23,14 +23,23 @@
...
             patch "finish"
           end
         end
       end
     end
     resources :course_subjects, only: :show
   end
+  resources :course_subjects, only: :show do
+    member do
+      patch "finish"
```
Comment (vanvtt-0952) [inline config/routes.rb]: dùng symbol dc ko bé

**Comment by lethihaivan** [comment]
ready

File: app/controllers/trainee/user_subjects_controller.rb (hunk 1 above)
Comment (vanvtt-0952) [reply app/controllers/trainee/user_subjects_controller.rb:37]: tìm hiển ensure đi bé nhé

**Comment by lethihaivan** [comment]
ready

File: app/controllers/trainee/user_subjects_controller.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
+class Trainee::UserSubjectsController < UsersController
+  before_action :load_user_subject, :correct_user
+
+  def finish
+    if @user_subject.joined? || @user_subject.finished?
+      flash[:danger] = t "user_subjects.not_active_finished"
```
Comment (vanvtt-0952) [inline app/controllers/trainee/user_subjects_controller.rb:7]: `return redirect_to trainee_course_subject_path @user_subject.course_subject`
xóa else

**Comment by lethihaivan** [comment]
ready

**Pull request merged**